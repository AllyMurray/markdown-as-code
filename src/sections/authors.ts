import { type ListOptions, ListSection } from './list-section.js';
import { link } from '../syntax/link.js';
import { mentionPerson } from '../syntax/mention.js';

interface Author {
  githubUsername: string;
}

export interface AuthorsOptions extends Partial<ListOptions> {}

export class Authors extends ListSection<Author> {
  constructor(options?: AuthorsOptions) {
    super({
      title: options?.title ?? 'Authors',
      type: 'Unordered',
      ...options,
    });
  }

  protected itemMapper(author: Author): string {
    return link(
      mentionPerson(author.githubUsername),
      new URL(author.githubUsername, 'https://www.github.com').toString()
    );
  }
}

export function authorsSection(options?: AuthorsOptions) {
  return new Authors(options);
}
