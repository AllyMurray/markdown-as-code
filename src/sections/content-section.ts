import { DocumentSection, DocumentSectionOptions } from './section.js';
import { heading } from '../syntax/heading.js';

interface ContentSectionOptions extends DocumentSectionOptions {
  content?: string;
}

export class ContentSection extends DocumentSection {
  private content: string;

  constructor(options: ContentSectionOptions) {
    super(options);
    this.content = options.content ?? '';
  }

  public appendContent(content: string) {
    this.content = this.content
      ? [this.content, '', content].join('\n')
      : content;
    return this;
  }

  protected synthesizeContent(): Array<string> {
    return [heading(this.headingLevel, this.title), '', this.content];
  }
}

export function contentSection(options: ContentSectionOptions) {
  return new ContentSection(options);
}
