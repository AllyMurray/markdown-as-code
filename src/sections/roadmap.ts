import { ListSection } from './list-section.js';

interface RoadmapItem {
  text: string;
}

export class Roadmap extends ListSection<RoadmapItem> {
  constructor(title?: string) {
    super(title ?? 'Roadmap', 'Unordered');
  }

  protected itemMapper(item: RoadmapItem): string {
    return item.text;
  }
}

export function roadmapSection(title?: string) {
  return new Roadmap(title);
}
