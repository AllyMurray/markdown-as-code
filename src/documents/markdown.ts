import { writeFileSync } from 'fs';
import { join } from 'path';
import { ContentSection } from '../sections/content-section.js';
import { DocumentSection, tryFindSection } from '../sections/section.js';
import { tableOfContentsSection } from '../sections/table-of-contents.js';
import { heading } from '../syntax/heading.js';

export interface MarkdownOptions {
  fileName: string;

  /**
   * The directory the markdown file will be written to
   *
   * @default project root directory
   */
  outDir?: string;

  title: string;
  description?: string;

  /**
   * Include a table of contents in the output
   *
   * @default true
   */
  tableOfContents?: boolean;
}

export interface SectionContent<Section extends DocumentSection> {
  updater: (section: Section) => void;
  path: Array<string>;
}

export class MarkdownDocument {
  private _outDir: string;
  private _tableOfContents: boolean;
  private _sections: Array<DocumentSection> = [];

  constructor(private options: MarkdownOptions) {
    this._outDir = options.outDir ?? process.cwd();
    this._tableOfContents = options.tableOfContents ?? true;
    return this;
  }

  public addSection(section: DocumentSection): this;
  public addSection(title: string, content: string): this;
  public addSection(
    sectionOrTitle: DocumentSection | string,
    content?: string
  ): this {
    if (sectionOrTitle instanceof DocumentSection) {
      this._sections.push(sectionOrTitle);
      return this;
    }
    if (!content) {
      throw new Error('Content is required');
    }
    this._sections.push(new ContentSection({ title: sectionOrTitle, content }));
    return this;
  }

  appendSection<Section extends DocumentSection>({
    path,
    updater,
  }: SectionContent<Section>) {
    const section = tryFindSection<Section>(this._sections, path);
    if (!section) {
      throw new Error(`Section not found: ${path.join(' > ')}`);
    }
    updater(section);
    return this;
  }

  private synthesizeSections(sections: Array<DocumentSection>) {
    const lines: Array<string> = [];
    for (const section of sections) {
      lines.push(section.synthesize());

      if (section.subSections.length > 0) {
        lines.push(...this.synthesizeSections(section.subSections));
      }
    }
    return lines;
  }

  public synthContent() {
    let tableOfContents;
    if (this._tableOfContents) {
      tableOfContents = tableOfContentsSection({
        sections: this._sections,
      }).synthesize();
    }

    const lines = [
      heading(1, this.options.title),
      this.options.description,
      tableOfContents,
      '',
    ];

    lines.push(...this.synthesizeSections(this._sections));

    return lines.filter(Boolean).join('\n\n');
  }

  public synth() {
    writeFileSync(
      join(this._outDir, this.options.fileName),
      this.synthContent()
    );
  }
}

export const createMarkdownDocument = (options: MarkdownOptions) =>
  new MarkdownDocument(options);
