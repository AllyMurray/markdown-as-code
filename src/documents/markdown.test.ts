import fs from 'fs';
import { createMarkdownDocument } from './markdown.js';
import { Authors, authorsSection } from '../sections/authors.js';
import { contentSection } from '../sections/content-section.js';
import { roadmapSection } from '../sections/roadmap.js';

vi.mock('fs');
const mockedFs = vi.mocked(fs);

describe('Markdown Document', () => {
  it('should synthesize a basic document', () => {
    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
    });

    expect(markdownDoc.synthContent()).toMatchSnapshot();
  });

  it('should synthesize a document after adding a section', () => {
    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
    }).addSection(roadmapSection({ items: [{ text: 'Item 1' }] }));

    expect(markdownDoc.synthContent()).toMatchSnapshot();
  });

  it('should synthesize a document after adding a custom section with a sub-section by setting rootSection', () => {
    const rootSection = contentSection({
      title: 'Root Section',
      content: 'Some markdown content',
    });

    contentSection({
      title: 'Sub Section',
      content: 'Some markdown sub content',
      parent: rootSection,
    });

    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
      content: [rootSection],
    });

    expect(markdownDoc.synthContent()).toMatchSnapshot();
  });

  it('should synthesize a document after adding a custom section with a sub-section by using addSubSection', () => {
    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
    }).addSection(
      contentSection({
        title: 'Root Section',
        content: 'Some markdown content',
      }).addSubSection(authorsSection().add({ githubUsername: 'jane-doe' })),
    );

    expect(markdownDoc.synthContent()).toMatchSnapshot();
  });

  it('should write a document to disk', () => {
    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
    });

    markdownDoc.synth();

    expect(mockedFs.writeFileSync.mock.calls[0][0]).toContain(
      'markdown-as-code/test.md',
    );
    expect(mockedFs.writeFileSync.mock.calls[0][1]).toMatchSnapshot();
  });

  it('should find a section by path', () => {
    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
    });

    markdownDoc.addSection(
      authorsSection({ items: [{ githubUsername: 'john-smith' }] }),
    );

    const authorsSectionFromPath = markdownDoc.tryFindSection(['Authors']);

    expect(authorsSectionFromPath).toBeInstanceOf(Authors);
  });

  it('should return undefined when section is not found by path', () => {
    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
    });

    markdownDoc.addSection(
      authorsSection({ items: [{ githubUsername: 'john-smith' }] }),
    );

    const authorsSectionFromPath = markdownDoc.tryFindSection(['Unknown']);

    expect(authorsSectionFromPath).toBeUndefined();
  });

  it('should return the file name', () => {
    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
    });

    expect(markdownDoc.fileName).toBe('test.md');
  });

  it('should return the out dir', () => {
    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
      outDir: 'lib',
    });

    expect(markdownDoc.outDir).toBe('lib');
  });

  it('should return the full file path', () => {
    const markdownDoc = createMarkdownDocument({
      title: 'Test',
      fileName: 'test.md',
      outDir: 'lib',
    });

    expect(markdownDoc.fullFilePath).toBe('lib/test.md');
  });
});
