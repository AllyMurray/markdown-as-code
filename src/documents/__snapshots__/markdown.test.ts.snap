// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Markdown Document > should synthesize a basic document 1`] = `
"# Test

## Table of Contents
"
`;

exports[`Markdown Document > should synthesize a document after adding a custom section with a sub-section 1`] = `
"# Test

## Table of Contents

- [Root Section](#root-section)
  - [Sub Section](#sub-section)

## Root Section

Some markdown content

### Sub Section

Some markdown sub content

Some more markdown sub content"
`;

exports[`Markdown Document > should synthesize a document after adding a custom section with a sub-section 2`] = `
"# Test

## Table of Contents

- [Acknowledgements](#acknowledgements)

## Acknowledgements

- [Some more markdown sub content](http://example.com)

- [Some more markdown sub content](http://example.com)"
`;

exports[`Markdown Document > should synthesize a document after adding a custom section with a sub-section 2 1`] = `
"# Test

## Table of Contents

- [Root Section](#root-section)
  - [Authors](#authors)

## Root Section

Some markdown content

### Authors

- [@jane-doe](https://www.github.com/jane-doe)

- [@john-smith](https://www.github.com/john-smith)"
`;

exports[`Markdown Document > should synthesize a document after adding a section 1`] = `
"# Test

## Table of Contents

- [Roadmap](#roadmap)

## Roadmap

- Item 1"
`;

exports[`Markdown Document > should write a document to disk 1`] = `
"# Test

## Table of Contents
"
`;
